File: src/middleware/errorHandler.ts
/**
 * @fileoverview Global Error Handler Middleware
 * @description Handles all errors in the application and sends standardized error responses
 * @module middleware/errorHandler
 */

import { Request, Response, NextFunction } from "express";
import { AppError } from "@/utils/AppError";
import { logger } from "@/utils/logger";
import { ApiResponse } from "@/types";
import { HTTP_STATUS } from "@/constants";

/**
 * Sensitive fields to redact from error logs
 * @constant {string[]} SENSITIVE_FIELDS
 */
const SENSITIVE_FIELDS = [
  "password",
  "otp",
  "token",
  "accessToken",
  "refreshToken",
  "secret",
  "apiKey",
  "apiSecret",
];

/**
 * Mask sensitive data in objects
 * @function maskSensitiveData
 * @description Recursively redacts sensitive fields from objects before logging
 * @param {any} obj - Object to mask
 * @returns {any} Object with sensitive fields redacted
 */
const maskSensitiveData = (obj: any): any => {
  try {
    // Deep clone the object to avoid mutating the original
    const clone = JSON.parse(JSON.stringify(obj || {}));

    /**
     * Recursive function to redact sensitive fields
     * @param {any} o - Object to process
     * @returns {any} Object with sensitive fields redacted
     */
    const redact = (o: any): any => {
      if (!o || typeof o !== "object") {
        return o;
      }

      // Process all keys in the object
      for (const key of Object.keys(o)) {
        // Redact sensitive fields
        if (SENSITIVE_FIELDS.includes(key.toLowerCase())) {
          o[key] = "[REDACTED]";
        }
        // Recursively process nested objects
        else if (o[key] && typeof o[key] === "object") {
          redact(o[key]);
        }
      }

      return o;
    };

    return redact(clone);
  } catch {
    // Return empty object if masking fails
    return {};
  }
};

/**
 * Handle Mongoose validation errors
 * @function handleValidationError
 * @param {any} error - Mongoose validation error
 * @returns {{ statusCode: number; message: string; errorType: string; errorText: string }}
 */
const handleValidationError = (
  error: any
): {
  statusCode: number;
  message: string;
  errorType: string;
  errorText: string;
} => {
  const errorMessages = Object.values(error.errors).map(
    (err: any) => err.message
  );

  return {
    statusCode: HTTP_STATUS.BAD_REQUEST,
    message: "Validation error",
    errorType: "Validation Error",
    errorText: errorMessages.join(", "),
  };
};

/**
 * Handle Mongoose duplicate key errors
 * @function handleDuplicateKeyError
 * @param {any} error - Mongoose duplicate key error
 * @returns {{ statusCode: number; message: string; errorType: string; errorText: string }}
 */
const handleDuplicateKeyError = (
  error: any
): {
  statusCode: number;
  message: string;
  errorType: string;
  errorText: string;
} => {
  const field = Object.keys(error.keyValue || {})[0] || "field";

  return {
    statusCode: HTTP_STATUS.CONFLICT,
    message: "Duplicate entry",
    errorType: "Duplicate Key Error",
    errorText: `${field} already exists`,
  };
};

/**
 * Handle JWT authentication errors
 * @function handleJWTError
 * @param {Error} error - JWT error
 * @returns {{ statusCode: number; message: string; errorType: string; errorText: string }}
 */
const handleJWTError = (
  error: Error
): {
  statusCode: number;
  message: string;
  errorType: string;
  errorText: string;
} => {
  if (error.name === "TokenExpiredError") {
    return {
      statusCode: HTTP_STATUS.UNAUTHORIZED,
      message: "Authentication failed",
      errorType: "Authentication Error",
      errorText: "Token expired",
    };
  }

  return {
    statusCode: HTTP_STATUS.UNAUTHORIZED,
    message: "Authentication failed",
    errorType: "Authentication Error",
    errorText: "Invalid token",
  };
};

/**
 * Global Error Handler Middleware
 * @function errorHandler
 * @description Centralized error handling middleware that processes all errors
 * and sends standardized error responses to the client
 * @param {Error} error - The error object
 * @param {Request} req - Express request object
 * @param {Response} res - Express response object
 * @param {NextFunction} next - Express next function
 * @returns {void}
 */
export const errorHandler = (
  error: Error,
  req: Request,
  res: Response,
  next: NextFunction
): void => {
  // Default error values
  let statusCode: number = HTTP_STATUS.INTERNAL_SERVER_ERROR;
  let message = "Internal Server Error";
  let errorType = "Server Error";
  let errorText: string | undefined;

  // Handle AppError instances (custom application errors)
  if (error instanceof AppError) {
    statusCode = error.statusCode;
    message = error.message;
    errorType = error.errorType || errorType;
    errorText = error.message; // Use the error message as error text
  }
  // Handle Mongoose validation errors
  else if (error.name === "ValidationError") {
    const errorInfo = handleValidationError(error as any);
    statusCode = errorInfo.statusCode;
    message = errorInfo.message;
    errorType = errorInfo.errorType;
    errorText = errorInfo.errorText;
  }
  // Handle Mongoose duplicate key errors (MongoDB error code 11000)
  else if ((error as any).code === 11000) {
    const errorInfo = handleDuplicateKeyError(error as any);
    statusCode = errorInfo.statusCode;
    message = errorInfo.message;
    errorType = errorInfo.errorType;
    errorText = errorInfo.errorText;
  }
  // Handle JWT errors
  else if (
    error.name === "JsonWebTokenError" ||
    error.name === "TokenExpiredError"
  ) {
    const errorInfo = handleJWTError(error);
    statusCode = errorInfo.statusCode;
    message = errorInfo.message;
    errorType = errorInfo.errorType;
    errorText = errorInfo.errorText;
  }
  // Handle CastError (invalid ObjectId format)
  else if (error.name === "CastError") {
    statusCode = HTTP_STATUS.BAD_REQUEST;
    message = "Validation error";
    errorType = "Validation Error";
    errorText = "Invalid ID format";
  }
  // Handle Multer errors (file upload errors)
  else if (error.name === "MulterError") {
    statusCode = HTTP_STATUS.BAD_REQUEST;
    message = "File upload error";
    errorType = "File Upload Error";

    const multerError = error as any;
    switch (multerError.code) {
      case "LIMIT_UNEXPECTED_FILE":
        errorText = "Only one file is allowed. Please upload a single file.";
        break;
      case "LIMIT_FILE_SIZE":
        errorText = `File size too large. Maximum allowed size is ${Math.round(
          (req as any).uploadLimit || 5
        )}MB.`;
        break;
      case "LIMIT_FILE_COUNT":
        errorText = "Too many files. Only one file is allowed.";
        break;
      case "LIMIT_FIELD_KEY":
        errorText = "Invalid field name for file upload.";
        break;
      default:
        errorText = multerError.message || "File upload failed.";
    }
  }
  // Handle other unexpected errors
  else {
    // Log unexpected errors for debugging
    logger.error("Unexpected error:", {
      name: error.name,
      message: error.message,
      stack: error.stack,
    });
  }

  // Log error with masked sensitive data
  logger.error("Error occurred:", {
    message: error.message,
    stack: error.stack,
    statusCode,
    errorType,
    error: errorText,
    url: req.url,
    method: req.method,
    ip: req.ip,
    userAgent: req.get("User-Agent"),
    body: maskSensitiveData(req.body),
    params: maskSensitiveData(req.params),
    query: maskSensitiveData(req.query),
  });

  // Send standardized error response
  const response: ApiResponse = {
    success: false,
    message,
    errorType: errorType || "Server Error",
    error: errorText || message,
    data: null,
  };

  res.status(statusCode).json(response);
};

File: app/middleware/error_handler.py
"""
Global exception handler middleware for FastAPI.
"""
import logging
import traceback
from typing import Callable

from fastapi import Request, Response, status
from fastapi.responses import JSONResponse
from pydantic import ValidationError as PydanticValidationError

from app.exceptions.errors import (
    BaseAppError,
    DatabaseConnectionError,
    DatabaseOperationError,
    OpenAIAPIError,
    OpenAIRateLimitError,
    OpenAITimeoutError,
    SessionNotFoundError,
)
from app.schemas.responses import ErrorResponse
from app.utils.error_handler import log_error_with_context

logger = logging.getLogger(__name__)


async def exception_handler_middleware(request: Request, call_next: Callable) -> Response:
    """
    Global exception handler middleware that catches all exceptions
    and returns standardized error responses.
    """
    try:
        response = await call_next(request)
        return response
    except Exception as exc:
        return await handle_exception(exc, request)


async def handle_exception(exc: Exception, request: Request) -> JSONResponse:
    """
    Handle exceptions and return appropriate HTTP responses.
    """
    # Log the error with context
    context = {
        "path": request.url.path,
        "method": request.method,
        "query_params": dict(request.query_params),
    }
    log_error_with_context(exc, context)
    
    # Handle custom application errors
    if isinstance(exc, SessionNotFoundError):
        return JSONResponse(
            status_code=status.HTTP_404_NOT_FOUND,
            content=ErrorResponse(
                status="error",
                message=exc.message,
                error_code=exc.error_code,
                details=exc.details
            ).model_dump()
        )
    
    if isinstance(exc, DatabaseConnectionError):
        return JSONResponse(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            content=ErrorResponse(
                status="error",
                message="Database service unavailable. Please try again later.",
                error_code=exc.error_code,
                details=exc.details
            ).model_dump()
        )
    
    if isinstance(exc, DatabaseOperationError):
        return JSONResponse(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            content=ErrorResponse(
                status="error",
                message="Database operation failed. Please try again later.",
                error_code=exc.error_code,
                details=exc.details
            ).model_dump()
        )
    
    if isinstance(exc, OpenAIRateLimitError):
        retry_after = exc.details.get("retry_after")
        headers = {}
        if retry_after:
            headers["Retry-After"] = str(retry_after)
        
        return JSONResponse(
            status_code=status.HTTP_429_TOO_MANY_REQUESTS,
            headers=headers,
            content=ErrorResponse(
                status="error",
                message="Service is temporarily unavailable due to high demand. Please try again later.",
                error_code=exc.error_code,
                details=exc.details
            ).model_dump()
        )
    
    if isinstance(exc, (OpenAITimeoutError, OpenAIAPIError)):
        return JSONResponse(
            status_code=status.HTTP_502_BAD_GATEWAY,
            content=ErrorResponse(
                status="error",
                message="AI service is temporarily unavailable. Please try again later.",
                error_code=exc.error_code,
                details=exc.details
            ).model_dump()
        )
    
    if isinstance(exc, BaseAppError):
        return JSONResponse(
            status_code=status.HTTP_400_BAD_REQUEST,
            content=ErrorResponse(
                status="error",
                message=exc.message,
                error_code=exc.error_code,
                details=exc.details
            ).model_dump()
        )
    
    # Handle Pydantic validation errors
    if isinstance(exc, PydanticValidationError):
        return JSONResponse(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            content=ErrorResponse(
                status="error",
                message="Validation error",
                error_code="VALIDATION_ERROR",
                details={"errors": exc.errors()}
            ).model_dump()
        )
    
    # Handle unexpected errors
    logger.error(
        f"Unhandled exception: {type(exc).__name__}: {str(exc)}",
        exc_info=True,
        extra={"path": request.url.path, "method": request.method}
    )
    
    return JSONResponse(
        status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
        content=ErrorResponse(
            status="error",
            message="An unexpected error occurred. Please try again later.",
            error_code="INTERNAL_SERVER_ERROR",
            details={}
        ).model_dump()
    )

